Add ConsoleCliDownloads controller

jhadvig opened the pull request:
https://jira.coreos.com/browse/CONSOLE-1423

Adding ConsoleCliDownloads controller that will take care of ConsoleCliDownloads `oc` binaries for different platforms.

/assign @benjaminapetersen

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,197 @@
+	MacPlatformKey     = "mac"
+	WindowsPlatformKey = "windows"
+)
+
```
nit: Probably we should caps `CLI` as `CLIDownloadsSyncController `.

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,197 @@
+}
+
+// factory func needs clients and informers
+func NewCliDownloadsSyncController(
```
nit: caps `CLI`.

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,197 @@
+	host := routesub.GetCanonicalHost(consoleRoute)
+	requiredOCCliDownloads := platformBasedCliDownloads(host, "amd64", ocCliDownloadsCustomResourceName)
+
+	_, _, cliDownloadsErr := GetOrCreate(c.consoleCliDownloadsClient, requiredOCCliDownloads)
```
Looks good so far, I bet we want to have a `HandleDegraded()` call here, with some potential reasons.

jhadvig replied:
unfortunatelly `HandleDegraded()` is a `consoleOperator` struct method, so can't use it here. We need either to refact the `status.go` methods to be able to call on different struct, or have a separate function for the `CLIDownloadsSyncController` struct

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,197 @@
+	}
+}
+
+func GetOrCreate(consoleClient consolev1client.ConsoleCLIDownloadInterface, required *v1.ConsoleCLIDownload) (*v1.ConsoleCLIDownload, bool, error) {
```
Contemplating, do we want a full `Apply` for this?  If it exists, but someone tampers with it, we probably want to stomp it back to the correct values.

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,197 @@
+		// clients
+		operatorConfigClient:      operatorClient,
+		consoleCliDownloadsClient: consoleClient,
+		routeClient:               routev1Client,
```
`consoleClient` as `cliDownloadsInterface`?

benjaminapetersen commented on pkg/console/starter/starter.go:
```diff
@@ -118,6 +128,11 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
 		oauthinformers.WithTweakListOptions(tweakListOptionsForOAuth),
 	)
 
+	consoleInformer := consoleinformers.NewSharedInformerFactory(
```
Prob plural like the others `consoleInformers`.

benjaminapetersen replied:
Then need to ensure we start the informers, this one is new:

```
	for _, informer := range []interface {
		Start(stopCh <-chan struct{})
	}{
                // etc.
		resourceSyncerInformers,
		operatorConfigInformers,
                // add here, and should be good!
		consoleInformers,
		configInformers,
                 // etc.
	} {
		informer.Start(ctx.Done())
	}
```

benjaminapetersen replied:
Yup, start the informer and you will get this:

![Screen Shot 2019-10-07 at 12 37 38 PM](https://user-images.githubusercontent.com/280512/66330914-5e6cbd80-e8ff-11e9-9ca9-afe4ecacc0da.png)

jhadvig commented:
@benjaminapetersen updated the PR. The controller will make sure the `ConsoleCliDownload` CR will be created for `oc` and `odo` (@spadgett FYI)
Already started to work on the unit tests but need to align on the format of the controller.

Also I've seen https://github.com/openshift/console-operator/pull/309 ... after it will land will update the PR with `HandleDegraded` as we spoke.

PTAL

benjaminapetersen commented on manifests/03-rbac-role-cluster-extensions.yaml:
```diff
@@ -9,10 +9,20 @@ rules:
   - get
   - list
   - watch
+- apiGroups:
```
Good call, this is better.

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,260 @@
+			DisplayName: "odo - Developer-focused CLI for OpenShift",
+			Links: []v1.Link{
+				{
+					Href: "https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/",
```
Are we good with this `href` being hard-coded?  I imagine it doesn't really change?  Any concerns about versions?

jhadvig replied:
good question, @spadgett do you know if there is any chance that this shall change ?
If so can add and entry to the operator config and read it from there ?

spadgett replied:
I'm not sure about odo API compatibility, but at some point, we might want to lock to a specific odo version for a specific openshift version. It is hard-coded to latest today.

@mohammedzee1000 @kadel We now have an extension point in console for adding to the command-line download page. Is there another operator that should own creating the CR for the odo download? If console owns it, what version should we link to?

kadel replied:
Currently, the odo is aiming to be compatible with 3.9 and up. odo is trying to be compatible with every supported OCP version. Using the latest should be fine.

There is no odo operator if the console can own this that would be great.

jhadvig replied:
@spadgett I guess than our operator will managed `odo` as well

spadgett replied:
@kadel Should we point to odo `latest`? Or a version like `v0.0.20`?

benjaminapetersen replied:
I'm fine managing it if we can point to `latest`.  If we get to a point where we need to keep up with version numbers, would we want to revisit the mechanism for that?

benjaminapetersen replied:
Also, just validating, `odo` is not built for several `platforms`?  `linux` only, I assume?  Should this be clarifiedin the text?

jhadvig replied:
> Also, just validating, odo is not built for several platforms? linux only, I assume? Should this be clarifiedin the text?

we probably should

kadel replied:
> Also, just validating, `odo` is not built for several `platforms`? `linux` only, I assume? Should this be clarifiedin the text?

odo is built for all 3 platforms (windows, mac, linux) all 3 of them are available at https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/

kadel replied:
> I'm fine managing it if we can point to `latest`. If we get to a point where we need to keep up with version numbers, would we want to revisit the mechanism for that?

Right now the only time where I see that we would want to keep up with the versions is when there are some big changes in core openshfit or k8s apis. Wouldn't that also result in OpenShift v5?

benjaminapetersen replied:
Should this then mirror the links we generate from console, by listing the 3 links separately?

```
odo-linux-amd64.tar.gz  
odo-darwin-amd64.tar.gz  
odo-windows-amd64.exe
```

benjaminapetersen replied:
![Screen Shot 2019-10-07 at 12 37 38 PM](https://user-images.githubusercontent.com/280512/66623814-68bae000-ebbb-11e9-859f-2e82a83dc2d0.png)

benjaminapetersen replied:
This would imply that the file names are stable, but is likely a better experience if we can confirm.

kadel replied:
I think that it might be safer to go with a link to the latest directory for now.
We are just preparing the 1.0 release, and it might change a little bit. After the release it will be stable, but I can't guarantee the exact filenames right now.

benjaminapetersen replied:
ok that works.

benjaminapetersen replied:
I would probably recommend the individual files as a better experience, once you are set.

benjaminapetersen commented on pkg/console/starter/starter.go:
```diff
@@ -118,6 +128,11 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
 		oauthinformers.WithTweakListOptions(tweakListOptionsForOAuth),
 	)
 
+	consoleInformers := consoleinformers.NewSharedInformerFactory(
```
I'm considering calling this `consoleExtensionInformers`, just for clarity.  Its hard to keep track of all the different configs.

jhadvig replied:
`consoleCLIDownloadsInformers`... per our conversation

benjaminapetersen replied:
👍

benjaminapetersen commented on pkg/console/starter/starter.go:
```diff
@@ -152,6 +167,15 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
 		resourceSyncer,
 	)
 
+	cliDownloadsController := clidownloads.NewCLIDownloadsSyncController(
```
Lets organize these with `client`/`informer` next to each other, `client` first like https://github.com/openshift/console-operator/pull/310/files#diff-69ecbb024dab972c8729678abdf9c0faR79.

I think it will help, granted .  don't think there will be much added to this.

```golang
cliDownloadsController := clidownloads.NewCLIDownloadsSyncController(
  // operator is to set status, correct?
  operatorConfigClient.OperatorV1(),
  operatorClient,
  // client, informer pair for ConsoleCLIDownloads
  consoleClient.ConsoleV1().ConsoleCLIDownloads(),
  consoleInformers.Console().V1().ConsoleCLIDownloads(),
  // route just to get route... but should we be informed of a route change?
  routesClient.RouteV1(),
  recorder,
)
```

benjaminapetersen replied:
I think we prob want the `routeInformer`.  If the `Host` changes on the route, then we need to update the `CLIDownloads`.

benjaminapetersen replied:
Check that, you'll need both for `route`:

```golang
routeClient // so we can GET the downloads route & pluck out the canonical host.
routeInformer // notify us of any route changes
```

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,261 @@
+
+func PlatformBasedOCConsoleCLIDownloads(host, arch, cliDownloadsName string) *v1.ConsoleCLIDownload {
+	baseURL := fmt.Sprintf("%s/%s", util.HTTPS(host), arch)
+
```
We could potentially have a map of strings here:

```golang
platforms := [string]string{
  "Linux": "linux",
   "Mac": "mac",
  "Windows": "windows",
}
```

and a loop to make the `Links`:

```golang
links := []v1.Link{}
for label, platform range platforms {
   links := append(links, Link{
      Href: GetPlatformURL(baseURL, platform, "oc.tar"),
      Text: fmt.Sprintf("Download oc for %v", Label)
   })
}
```
That or maybe just inline the `*PlatformKey`s since they are tied to the specifc "Download for *Platform*` text.

benjaminapetersen replied:
Nice!

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,261 @@
+		},
+	}
+}
+
```
Pondering if we should make this 

```golang
ApplyConsoleCLIDownload(....)
``` 
We can probably look at a func in `library-go` like:

```golang
func ApplyService(client coreclientv1.ServicesGetter, recorder events.Recorder, required *corev1.Service) (*corev1.Service, bool, error) {
  // does things
}
```
and copy it.  I imagine we want the `eventRecorder` in there to create events if things go awry. 

So something like:

```golang
func ApplyConsoleCLIDownload(cliDownloadsClient CliDownloadsGetter, recorder events.Recorder, required *v1.ConsoleCLIDownload) (v1.ConsoleCLIDownload, string, error) {
  // more or less copy the ApplyService func, but tweak it to fit this resource.
}
```

jhadvig replied:
`CreateOrUpdate` already has similar logic... need to add the recored and rename though :)

jhadvig replied:
so looked at adding the recorder but noticed that the methods that the recorder is passed into are private (`reportCreateEvent` && `reportUpdateEvent`)

Also our custom apply [CustomApplyOAuth](https://github.com/openshift/console-operator/blob/master/pkg/console/subresource/oauthclient/oauthclient.go#L21) method is not using the recored as well, supposedly due to the fact the methods are private.

Would ship it for now and add this as a techDebt ?

benjaminapetersen replied:
Agree!

Do you want to throw a couple `TODO` comments in there quick?  This would be helpful if we promote it to `openshift/library-go/pkg/console/resource/resourceapply/core.go` (does not yet exist, but if others are going to use these in their operators, we could ensure the appropriate `apply` funcs exist for each of our custom resources.)

jhadvig commented:
@benjaminapetersen Ive updated the PR based on comments also added comments.

 Added a `removeCLIDownloads()` method for the `Removed` operator state.

Only thing that is missing are the e2e tests that I've started to work on, but I think it would be better to address them as a separate PR, since this is already quite beefy 😄 

Other thing is that we need to figure out https://github.com/openshift/console-operator/pull/306#discussion_r333060441

PTAL

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,285 @@
+	if err != nil {
+		return err
+	}
+
```
Pondering if we want 2 different `types`:

```golang
status.HandleDegraded(client, "OCDownloadsSync", ....)
status.HandleDegraded(client, "ODODownloadsSync",....)
```
Since we `return err` immediately after, we can't really stomp over the `oc` sync error with an `odo` sync error. But, in general, I think we don't want to reuse `type` to avoid the potential for stomping on an err (say if we refactor down the road and the shape here changes).

jhadvig replied:
Makes sense to me !

jhadvig commented:
/retest

jhadvig commented:
/retest

jhadvig commented:
/retest

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,295 @@
+	consoleCliDownloadsClient consoleclientv1.ConsoleCLIDownloadInterface
+	routeClient               routeclientv1.RoutesGetter
+	// listers
+	routesLister         routeslisterv1.RouteLister
```
So you can drop the `routesLister` now, we aren't using it.

Recalling that:
- A client handles CRUD in our get,list,watch,create,update,delete, etc
- a Lister is a `cached` version of `Get` and `List`.  We don't really need it (and we aren't actually implementing it anyway here) since there are not a lot of these resources.

jhadvig commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,295 @@
+	operatorClient.Informer().AddEventHandler(ctrl.newEventHandler())
+	operatorConfigInformer.Informer().AddEventHandler(ctrl.newEventHandler())
+	consoleCLIDownloadsInformers.Informer().AddEventHandler(ctrl.newEventHandler())
+	routesInformers.Informer().AddEventHandler(ctrl.newEventHandler())
```
> So you can drop the routesLister now, we aren't using it.

shouldn't we keep it at least here ?

benjaminapetersen replied:
So, you do need the `informer`, yes!
And the `client`.
But we don't need the `lister`.

jhadvig commented:
@benjaminapetersen comments addressed. Left the `odo` link to the latest directory, instead of particular platform binary, per [discussion](https://github.com/openshift/console-operator/pull/306#discussion_r334367942) with @kadel 

PTAL

jhadvig commented:
/retest

jhadvig commented:
/retest

benjaminapetersen commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -0,0 +1,294 @@
+		// clients
+		consoleCliDownloadsClient: cliDownloadsInterface,
+		routeClient:               routeClient,
+		// listers
```
You don't need the `operatorconfiglister` here.  We only need the `informer` below to `Inform()` & kick our watches.  And we only need a `client` for when we need to do a `.Get()`.

benjaminapetersen replied:
After a little more research and a few more conversations, confirmed we only need the `Informer` and `client` for the actual `CLIDownloads` resource.  We shouldn't watch operator config or anything else.

benjaminapetersen commented:
/lgtm

benjaminapetersen commented:
/retest

Terraform failed.

benjaminapetersen commented:
And here is our test failure:

```
fail [github.com/openshift/origin/test/extended/authorization/rbac/groups_default_rules.go:221]: Oct 15 19:10:38.376: system:authenticated has extra permissions in namespace "":
{APIGroups:["console.openshift.io"], Resources:["consoleclidownloads"], Verbs:["create" "update" "delete"]}
```

@spadgett

benjaminapetersen commented:
/retest

https://github.com/openshift/origin/pull/23975 has merged so test should be unblocked.

jhadvig commented:
/retest

jhadvig commented:
/retest

jhadvig commented:
/retest

jhadvig commented:
/retest

spadgett commented on manifests/03-rbac-role-cluster-extensions.yaml:
```diff
@@ -9,10 +9,20 @@ rules:
+  - get
+  - list
+  - watch
+  - create
```
This should only be get/list/watch

benjaminapetersen replied:
I think he needs full `CRUD` for this one, since the `operator` is creating 2 of this resource, or updating if tampered with.

benjaminapetersen replied:
I updated the test incorrectly, added this resource to the same list as the rest of our resources.  My mistake, I forgot this resource is more actively managed than our others.

benjaminapetersen replied:
So you probably need to pull the `consoleclidownloads` out of `ClusterRole` `console-extensions-reader` file and put it in its own file. Name the role something like `console-clidownloads-extension-manager` and then create a `ClusterRoleBinding` for `console-operator`.

Then we can update the `origin` test to allow this.

spadgett replied:
No, this is wrong. This is the role that gets assigned to `system:authenticated`. We *cannot* let all users edit these. The operator role is a different one.

spadgett replied:
> I updated the test incorrectly, added this resource to the same list as the rest of our resources. My mistake, I forgot this resource is more actively managed than our others.

Your origin test updates are correct.

benjaminapetersen replied:
Right, intended to mean "console operator only" definitely not, "everyone".

spadgett commented:
/lgtm cancel
RBAC is incorrect

jhadvig commented:
@benjaminapetersen PR updated. PTAL

benjaminapetersen commented:
/lgtm
